### PR TITLE
Async serial connection

### PR DESCRIPTION
**Pull request by Hellowlol**

### CODE REVIEW

File: custom_components/ams/__init__.py [hunk 1]
```diff
@@ -21,26 +22,30 @@
...
     hass.data[DOMAIN] = hub
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, hub.stop_serial_read())
     hass.async_add_job(hass.config_entries.async_forward_entry_setup(entry, "sensor"))
+    _LOGGER.debug("async_setup_entry end")
     return True
 
 
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload a config entry."""
+    # What the diff from unload and remove..
```
Comment (turbokongen) [inline custom_components/ams/__init__.py:41]: https://developers.home-assistant.io/docs/config_entries_index/#unloading-entries

**Comment by Hellowlol** [comment]
Seems to be working, updated to current master. Anybody that can test it? It would be nice if somebody also compared the memory and CPU usage before and after this just to make sure that async don't preform worse then the threaded version.

Delete the integration, all the __pycache__ folder and setup up, again. Make sure you test both auto and the name of you meter manufacturer in the settings page.

**Comment by turbokongen** [comment]
Config is broken with parser setup.
I have kamstrup set, but it selects kaifa.
Also closing homeassistant throws error.
```
2020-04-13 09:38:10 WARNING (MainThread) [homeassistant.loader] You are using a custom integration for ams which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant.
2020-04-13 09:38:17 INFO (MainThread) [homeassistant.setup] Setting up ams
2020-04-13 09:38:17 INFO (MainThread) [homeassistant.setup] Setup of domain ams took 0.0 seconds.
2020-04-13 09:38:17 DEBUG (MainThread) [custom_components.ams] Connecting to HAN using port /dev/ttyUSB0
2020-04-13 09:38:17 DEBUG (MainThread) [custom_components.ams] DUMP entry {'meter_manufacturer': 'kamstrup', 'parity': 'N', 'serial_port': '/dev/ttyUSB0'}
2020-04-13 09:38:17 DEBUG (MainThread) [custom_components.ams] DUMP entry {'meter_manufacturer': 'kamstrup', 'parity': 'N', 'serial_port': '/dev/ttyUSB0'}
2020-04-13 09:38:17 DEBUG (MainThread) [custom_components.ams] Finish init of AMS
2020-04-13 09:38:17 DEBUG (MainThread) [custom_components.ams] Finish init of AMS
2020-04-13 09:38:17 DEBUG (MainThread) [custom_components.ams] Setting up serial.
2020-04-13 09:38:18 INFO (MainThread) [homeassistant.components.sensor] Setting up sensor.ams
2020-04-13 09:38:19 DEBUG (MainThread) [custom_components.ams] Using parser kamstrup
2020-04-13 09:38:19 DEBUG (MainThread) [custom_components.ams] Using parser kamstrup
2020-04-13 09:38:21 DEBUG (MainThread) [custom_components.ams.parsers.kaifa] Invalid packet size 228
2020-04-13 09:38:21 DEBUG (MainThread) [custom_components.ams] failed package: [126, 160, 226, 43, 33, 19, 35, 154, 230, 231, 0, 15, 0, 0, 0, 0, 12, 7, 228, 4, 13, 1, 9, 38, 20, 255, 128, 0, 0, 2, 25, 10, 14, 75, 97, 109, 115, 116, 114, 117, 112, 95, 86, 48, 48, 48, 49, 9, 6, 1, 1, 0, 0, 5, 255, 10, 16, 53, 55, 48, 54, 53, 54, 55, 50, 55, 49, 53, 56, 57, 52, 49, 57, 9, 6, 1, 1, 96, 1, 1, 255, 10, 18, 54, 56, 52, 49, 49, 50, 49, 66, 78, 50, 52, 51, 49, 48, 49, 48, 52, 48, 9, 6, 1, 1, 1, 7, 0, 255, 6, 0, 0, 8, 84, 9, 6, 1, 1, 2, 7, 0, 255, 6, 0, 0, 0, 0, 9, 6, 1, 1, 3, 7, 0, 255, 6, 0, 0, 0, 0, 9, 6, 1, 1, 4, 7, 0, 255, 6, 0, 0, 0, 85, 9, 6, 1, 1, 31, 7, 0, 255, 6, 0, 0, 1, 32, 9, 6, 1, 1, 51, 7, 0, 255, 6, 0, 0, 2, 165, 9, 6, 1, 1, 71, 7, 0, 255, 6, 0, 0, 3, 61, 9, 6, 1, 1, 32, 7, 0, 255, 18, 0, 235, 9, 6, 1, 1, 52, 7, 0, 255, 18, 0, 237, 9, 6, 1, 1, 72, 7, 0, 255, 18, 0, 234, 50, 78, 126]
2020-04-13 09:38:31 DEBUG (MainThread) [custom_components.ams.parsers.kaifa] Invalid packet size 228
2020-04-13 09:38:31 DEBUG (MainThread) [custom_components.ams] failed package: [126, 160, 226, 43, 33, 19, 35, 154, 230, 231, 0, 15, 0, 0, 0, 0, 12, 7, 228, 4, 13, 1, 9, 38, 30, 255, 128, 0, 0, 2, 25, 10, 14, 75, 97, 109, 115, 116, 114, 117, 112, 95, 86, 48, 48, 48, 49, 9, 6, 1, 1, 0, 0, 5, 255, 10, 16, 53, 55, 48, 54, 53, 54, 55, 50, 55, 49, 53, 56, 57, 52, 49, 57, 9, 6, 1, 1, 96, 1, 1, 255, 10, 18, 54, 56, 52, 49, 49, 50, 49, 66, 78, 50, 52, 51, 49, 48, 49, 48, 52, 48, 9, 6, 1, 1, 1, 7, 0, 255, 6, 0, 0, 8, 96, 9, 6, 1, 1, 2, 7, 0, 255, 6, 0, 0, 0, 0, 9, 6, 1, 1, 3, 7, 0, 255, 6, 0, 0, 0, 0, 9, 6, 1, 1, 4, 7, 0, 255, 6, 0, 0, 0, 64, 9, 6, 1, 1, 31, 7, 0, 255, 6, 0, 0, 1, 31, 9, 6, 1, 1, 51, 7, 0, 255, 6, 0, 0, 2, 166, 9, 6, 1, 1, 71, 7, 0, 255, 6, 0, 0, 3, 68, 9, 6, 1, 1, 32, 7, 0, 255, 18, 0, 235, 9, 6, 1, 1, 52, 7, 0, 255, 18, 0, 237, 9, 6, 1, 1, 72, 7, 0, 255, 18, 0, 234, 226, 219, 126]
2020-04-13 09:38:41 DEBUG (MainThread) [custom_components.ams.parsers.kaifa] Invalid packet size 228
2020-04-13 09:38:41 DEBUG (MainThread) [custom_components.ams] failed package: [126, 160, 226, 43, 33, 19, 35, 154, 230, 231, 0, 15, 0, 0, 0, 0, 12, 7, 228, 4, 13, 1, 9, 38, 40, 255, 128, 0, 0, 2, 25, 10, 14, 75, 97, 109, 115, 116, 114, 117, 112, 95, 86, 48, 48, 48, 49, 9, 6, 1, 1, 0, 0, 5, 255, 10, 16, 53, 55, 48, 54, 53, 54, 55, 50, 55, 49, 53, 56, 57, 52, 49, 57, 9, 6, 1, 1, 96, 1, 1, 255, 10, 18, 54, 56, 52, 49, 49, 50, 49, 66, 78, 50, 52, 51, 49, 48, 49, 48, 52, 48, 9, 6, 1, 1, 1, 7, 0, 255, 6, 0, 0, 8, 88, 9, 6, 1, 1, 2, 7, 0, 255, 6, 0, 0, 0, 0, 9, 6, 1, 1, 3, 7, 0, 255, 6, 0, 0, 0, 0, 9, 6, 1, 1, 4, 7, 0, 255, 6, 0, 0, 0, 97, 9, 6, 1, 1, 31, 7, 0, 255, 6, 0, 0, 1, 32, 9, 6, 1, 1, 51, 7, 0, 255, 6, 0, 0, 2, 163, 9, 6, 1, 1, 71, 7, 0, 255, 6, 0, 0, 3, 61, 9, 6, 1, 1, 32, 7, 0, 255, 18, 0, 235, 9, 6, 1, 1, 52, 7, 0, 255, 18, 0, 237, 9, 6, 1, 1, 72, 7, 0, 255, 18, 0, 234, 233, 4, 126]
2020-04-13 09:38:44 ERROR (MainThread) [custom_components.ams] Some crap happend 
  File "/home/pi/.homeassistant/custom_components/ams/__init__.py", line 95, in aio_run
  File "/home/pi/.homeassistant/custom_components/ams/__init__.py", line 129, in read_bytes
  File "/usr/lib/python3.7/asyncio/streams.py", line 640, in read
  File "/usr/lib/python3.7/asyncio/streams.py", line 473, in _wait_for_data
  File "/home/pi/.homeassistant/custom_components/ams/__init__.py", line 117, in stop_serial_read
```

**Comment by Hellowlol** [comment]
I forgot the test non auto version. I have prob picked the wrong sequence. Can you try with auto?

Sendt fra min iPhone

File: custom_components/ams/parsers/__init__.py [hunk 2]
```diff
@@ -1,26 +1,25 @@
...
-def field_type(default="", fields=None, enc=str, dec=None):
-    """Obis/data field decoder/encoder."""
-    data = default.join(enc(i) for i in fields)
-    if dec:
-        return dec(data)
-    return data
+def auto_detect_parser(pkg):
+    """Helper to detect meter manufacturer."""
+    if pkg == "kamstrup":
+        pkg = KAIFA_METER_SEQ
```
Comment (Hellowlol) [inline custom_components/ams/parsers/__init__.py:8]: Zomg.

**Comment by turbokongen** [comment]
Ok, here are the result on my pi3+ of 1 hour running each.
Left side of graph is master branch and right side is this PR.
Spike in the left is the pi running hourly cronjob of rngd test.
Skjermbilde
Skjermbilde-1
Skjermbilde-2
Skjermbilde-3

**Comment by turbokongen** [comment]
Still a hickup at shutdown:
```
2020-04-13 11:08:55 INFO (MainThread) [homeassistant.components.websocket_api.http.connection.1745993296] Connection closed by client
2020-04-13 11:08:55 ERROR (MainThread) [custom_components.ams] Some crap happend 
Traceback (most recent call last):
  File "/home/pi/.homeassistant/custom_components/ams/__init__.py", line 95, in aio_run
    data = await self.read_bytes()
  File "/home/pi/.homeassistant/custom_components/ams/__init__.py", line 129, in read_bytes
    buf = await self._ser.read(1)
  File "/usr/lib/python3.7/asyncio/streams.py", line 640, in read
    await self._wait_for_data('read')
  File "/usr/lib/python3.7/asyncio/streams.py", line 473, in _wait_for_data
    await self._waiter
concurrent.futures._base.CancelledError
2020-04-13 11:09:00 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/home/pi/.homeassistant/custom_components/ams/__init__.py", line 117, in stop_serial_read
    self._ser.close()
AttributeError: 'StreamReader' object has no attribute 'close'
```

**Comment by Hellowlol** [comment]
Is was supposed to close the transport. I don’t think this is needed. I’ll remove this.

**Comment by Hellowlol** [comment]
I have fixed the bugs you mentioned.

**Comment by Hellowlol** [comment]
@jnrksv Can you test this PR=

File: custom_components/ams/__init__.py [hunk 3]
```diff
@@ -40,50 +39,95 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
...
+        try:
+            while self._running:
+                data = await self.read_bytes()
+                if parser.test_valid_data(data):
+                    _LOGGER.debug("data read from port=%s", data)
+                    self.sensor_data, _ = parser.parse_data(self.sensor_data, data)
+                    self._check_for_new_sensors_and_update(self.sensor_data)
+                else:
+                    _LOGGER.debug("failed package: %s", data)
+        except Exception as e:
```
Comment (turbokongen) [inline custom_components/ams/__init__.py:102]: Need to catch `asyncio.CancelledError` for shutdown.

**Comment by jnrksv** [comment]
Works fine on my system. Setup using "auto" worked. No increase in cpu usage. Just a slight increase in memory use (1-1.25% as shown below).
pr27

**Comment by Hellowlol** [comment]
I'm busy with plexapi atm, ill come back to this later.

**Comment by dkcsn** [comment]
Hi - and what an awesome plugin.
This module is referered as working
[(https://www.aliexpress.com/item/32834331647.html?spm=a2g0o.detail.1000060.1.74cfdcd4qts4jp)]

But for the yaml config - USB seems to be the only way?
ams:
  serial_port: '/dev/ttyUSB0' # Required
And the aliexpress module has no USB - so how to do that in the yaml?
I have one of the supported Kamstrup and have ordered the IR eye reader, so it would be great to know the idea of the non USB module?

**Pull request closed without merging**